addons/mail/tests/discuss/test_discuss_channel.py

```diff
@@ -546,45 +540,61 @@ def test_channel_private_unfollow(self):
-        The message should only be posted when the partner is still a member of the channel
+        When a partner leaves a group, the system will help post a message under
+        that partner's name in the group to notify others if `email_sent` is set `False`.
+        The message should only be posted when the partner is still a member of the group
         before method `_action_unfollow()` is called.
         If the partner has been removed earlier, no more messages will be posted
         even if `_action_unfollow()` is called again.
         '''
-        channel = self.env['discuss.channel'].browse(self.test_channel.id)
-        channel._add_members(partners=self.test_partner)
+        test_group = self.env['discuss.channel'].create({
+            'name': 'Private Channel',
+            'channel_type': 'group',
+        })
+        test_group._add_members(partners=self.test_partner)
 
         # no message should be posted under test_partner's name
         messages_0 = self.env['mail.message'].search([
             ('model', '=', 'discuss.channel'),
-            ('res_id', '=', channel.id),
+            ('res_id', '=', test_group.id),
             ('author_id', '=', self.test_partner.id)
         ])
-        self.assertEqual(len(messages_0), 0)
+        self.assertFalse(messages_0)
 
         # a message should be posted to notify others when a partner is about to leave
-        channel._action_unfollow(self.test_partner)
+        test_group._action_unfollow(self.test_partner)
         messages_1 = self.env['mail.message'].search([
             ('model', '=', 'discuss.channel'),
-            ('res_id', '=', channel.id),
+            ('res_id', '=', test_group.id),
             ('author_id', '=', self.test_partner.id)
         ])
         self.assertEqual(len(messages_1), 1)
 
         # no more messages should be posted if the partner has been removed before.
-        channel._action_unfollow(self.test_partner)
+        test_group._action_unfollow(self.test_partner)
         messages_2 = self.env['mail.message'].search([
             ('model', '=', 'discuss.channel'),
-            ('res_id', '=', channel.id),
+            ('res_id', '=', test_group.id),
             ('author_id', '=', self.test_partner.id)
         ])
         self.assertEqual(len(messages_2), 1)
         self.assertEqual(messages_1, messages_2)
 
+    def test_channel_join_unfollow_should_not_post_message(self):
+        channel = self.env['discuss.channel'].browse(self.test_channel.id)
+        channel._add_members(partners=self.test_partner)
+
+        # no message should be posted to notify others when a partner is joined and left
+        channel._action_unfollow(self.test_partner)
+        messages = self.env['mail.message'].search([
+            ('model', '=', 'discuss.channel'),
+            ('res_id', '=', channel.id),
+            ('author_id', '=', self.test_partner.id)
+        ])
+        self.assertFalse(messages)
```

Comment:
The test doesn't actually cover the "join" message, as the added member is done by a different user, so the author domain is not correct. This is the typical problem with negative tests...

It can be fixed for example by adding `.with_user(self.test_user)` before the `_add_members`